week-1.py

```diff
@@ -0,0 +1,75 @@
+import unittest
+
+# the algo that I have used is similar to Kadane Algorithm
+
+def longest_substring(string):
+
+
+	if(string is None):
+		return 0
+
+	L=len(string)
+
+	### stores the final answer
+	max_l=1
+
+	## stores the length of the current substring
+	curr_l=1
+
+	#### store the previous recent index value of all the characters 
+	visited_index=255*[-1] 
+	i=0
+
+	visited_index[ord(string[i])]=i
+	i=i+1
+
+	#traverse the whole string 
+	while(i<L):
+	    
+	    k=ord(string[i])
+	    prev_index=visited_index[k]
+
+	    
+	    #### if a new char is found that is not present in current substring 
+	    if((prev_index==-1) or (prev_index<i-curr_l)): 

```

Comment:
I am a bit confused by the 2nd clause of your if statement here, can you explain it for me, or write a test case where it gets used?

Reply:
Yes sure. Will be happy to explain my logic. 
Suppose the input string is 'geeksforgeeks'. 
1. The algo starts constructing the substrings starting from 'g'.
     It reaches till 'ge'-  curr_l =2.
2. After this, it encounters an 'e'.  So max_l value gets updated-> max_l=2. 
3. Now the current substring gets updated and it starts from the 2nd 'e'.
4. The code goes on adding letters to this current substring till it becomes 'eksfor'.
5 Then it encounters a 'g' again. Here is when the second clause comes into play. The clause checks whether the previous index of 'g' is within the length of the current substring. Since its not in this case, it adds this 'g' to the current substring. So it becomes- 'eksforg' with curr_l=7

Reply:
Thanks!